Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,200 @@
+.. SPDX-License-Identifier: GPL-2.0-only
+
+===========
+Ancillary Bus
+===========
+
+See <linux/ancillary_bus.h> for the models for ancillary_device and
+ancillary_driver.
+
+In some subsystems, the functionality of the core device (PCI/ACPI/other) may
+be too complex for a single device to be managed as a monolithic block or
+a part of the functionality might need to be exposed to a different subsystem.
+Splitting the functionality into smaller orthogonal devices would make it
+easier to manage data, power management and domain-specific communication with
+the hardware. A key requirement for such a split is that there is no dependency
+on a physical bus, device, register accesses or regmap support. These
+individual devices split from the core cannot live on the platform bus as they
+are not physical devices that are controlled by DT/ACPI. The same argument
+applies for not using MFD in this scenario as it relies on individual function
+devices being physical devices that are DT enumerated.
+
+An example for this kind of requirement is the audio subsystem where a single
+IP handles multiple entities such as HDMI, Soundwire, local devices such as
+mics/speakers etc. The split for the core's functionality can be arbitrary or
+be defined by the DSP firmware topology and include hooks for test/debug. This
+allows for the audio core device to be minimal and tightly coupled with
+handling the hardware-specific logic and communication.
+
+The ancillary bus is intended to be minimal, generic and avoid domain-specific
+assumptions. Each ancillary_device represents a part of its parent
+functionality. The generic behavior can be extended and specialized as needed
+by encapsulating an ancillary_device within other domain-specific structures and
+the use of .ops callbacks. Devices on the same ancillary bus do not share any
+structures and the use of a communication channel with the parent is
+domain-specific.
+
+When Should the Ancillary Bus Be Used
+=====================================
+
+When a kernel driver registers either an ancillary_device or ancillary_driver on
+the ancillary bus, we can use the nomenclature to refer to this kernel driver as
```

Comment:
the ancillary_driver may be registered in a separate module, completely independent from the registering driver.

This should be:

When a kernel driver registers either an ancillary_device ~~or ancillary_driver~~ on the ancillary bus

Reply:
No, the ancillary_driver is the stuct ancillary_driver { } that is *registered* on the bus.  

In your host/client nomenclature the host drivers are the registering drivers and the clients are the ancillary_driver and ancillary_device.  I was using term parent before and that was problematical for at least one reviewer.  I don't see how host is any better.  I settled on *registering driver* as that is exactly what the parent driver does.  It registers the object (ancillary_device or ancillary_driver).  So, whichever kernel object actually performs the call to register an object on the ancillary bus is the registering driver - it can only be registered once, so there is only one registering driver for each object on the anciallry bus, although a single registering driver can be the registering driver for multiple objects.

The module that is registering an object on the bus is responsible for the memory for that object for its lifespan.

Reply:
ok.

We are using verbiage in the introduction that a core device is being being split into orthogonal ancillary devices. So a core driver that carves/registers ancillary device(s) on the bus and an ancillary driver that binds to the ancillary device(s) seems more befitting to me.

Seems like you tried similar nomenclature but was not received well? If others are ok with this 'registering driver' nomenclature which denotes the driver that registers the ancillary_device or the ancillary_driver object on the bus, then move on.

Reply:
@dmertman I have to agree with @plbossart the ancillary driver does not need to registered by a kernel driver. In our SOF case, it is registered as a separate module.

Reply:
Maybe the confusion here is around the word driver.  What I am trying to say, is whatever kernel module performs the ancillary_register_driver() call is the registering module/driver/kernel object/etc.  And that module is responsible for the allocating the memory for what it is registering on the ancillary bus.

I can change the documentation to use the term "registering module" if that would be clearer?

Reply:
re-reading the initial proposal, the confusion comes from using 'registering driver' for two completely different purposes. One is device registration and the other is driver registration. Even 'registering module' is confusing since it doesn't clarify if this is the module which will register the device or the one which will register the driver.

Reply:
There are two registering drivers - one for the ancillary_device and one for the ancillary_driver.  A registering driver is a kernel driver/object/etc that registers *something* on the ancillary bus - be it a driver or a device.

Reply:
ok, but then the following paragraph is very strange, you use 'registering drivers' with two different meanings in the same sentence:

````
The ancillary bus is to be used when multiple registering drivers, who share a
common header file, need a mechanism to connect and provide access to a shared
object allocated by the ancillary_device's registering driver. 
````

In the first sentence, the 'multiple registering drivers' refers to a module which registers an ancillary_driver while the latter refers to the registration of an ancillary device.

Maybe change the last part to say 'to be used when multiple modules or subsystems register ancillary_drivers and need a  mechanism to connect and provide access to a shared object allocated during the registration of an ancillary device'.

````
 The registering
drivers for the ancillary_devices and ancillary_drivers can be from the same
subsystem, or from multiple subsystems.
````

That part is fine but can be moved first to help explain the 'multiple' part.